### PR TITLE
Issue 21

### LINKED ISSUE #21: Tray and expanded order management on the Database
We need to introduce more functionality when operating on the database and managing trays. We need to introduce the following three methods for tray management: createTray(Integer), veiwTrays() and removeFromTray(Order). Finally, for order management, we will need to introduce two new methods as well. fetchOrders() and completeOrder(Integer). Finally, we will need to create a Tray object type.

Datatypes:
-----
Tray:
Will contain the following information as attributes. Could also seperate the Tray wrapper into two seperate datatypes as trays that are baking will have different attributes from ones that have finished.
  1. id unique to the tray: String
  2. quantity: Integer
  3. donut id: Integer
  4. startBakingTime: String
  5. endBakingTime: String
  6. expireTime: String
-----
Methods:
-----

createTray(Integer): boolean
- This method will create a tray of a given donut type.
- A tray has the following parameters it must fill in the database:
  1. id unique to the tray (We can key the database and have it auto generate this for us).
  2. quantity set to 20.
  3. donut id.
  4. startBakingTime this will be the time when the donut was put in the oven.
  5. endBakingTime this will be the time of completion 2 hours later.
  6. expireTime this will be two days after the date we've started baking.
-----
viewTrays(): ArrayList<Tray>
- This will perform a select on the database for trays that have not yet expired.
- Return an Array List of tray objects. The tray object will contain the database attributes for each tray within the database.
-----
removeFromTray(Order): boolean
- This first call viewTrays(). Using the array list it will subtract from the quantity within each Tray until one of the two scenarios:
  1. Order has been satisfied.
  2. Order does not have enough donuts to satisfy the request.
    - The program will auto generate the required trays to fulfill the order. It is assumed the employees are in the process of satisfying the requirements or already have.
-----
fetchOrders(): ArrayList<Cart>
- Return an ArrayList of carts for every order that is currently marked incomplete.
- Each cart object encapsulates an ArrayList of Orders unique to an order.
  - TODO update cart to have an id parameter which will only be filled upon fetching.
-----
completeOrder(Integer): boolean
- Takes in a given integer and will mark every order with the given id as completed.
  - Assumes that each order is completed within the same day.
- Calls removeFromTray() for each donut entry to update inventory as required.
-----
completeOrder(Integer, String): boolean
- Takes in a given integer which is the order id.
- Takes in a given String which will represent a date in YYYY-MM-DD format.
- Calls removeFromTray() for each donut entry to update inventory as needed.
-----

### PR DESCRIPTION
**Pull request by Korbrent**
Implement a `Tray` class for managing inventory.
`Tray` methods: **All of these are STATIC methods**
- `Tray.getBakingTrays()`: Gets all trays that are currently in the oven.
- `Tray.getInventoryTrays()`: Gets all trays in the inventory that **are not** stale.
- `Tray.getStaleTrays()`: Gets all the trays in the inventory that **are** stale.
- `Tray.updateBakedTrays()`: Converts all trays from being `BakingTray` to `InventoryTray` trays. This should be called by the baker after the baked trays have been removed from the oven.
- `Tray.updateStaleTrays()`: Removes all stale trays from the system. This should be called by the baker after the stale donuts have been discarded
 
- `Tray.getTrays()`: Returns all trays. I doubt you need this.

All `Tray` objects are either an instanceof `BakingTray` or `InventoryTray`. 

**All of these methods are object methods**
`BakingTray`
- `obj.isBaking()`: returns true if the tray is still baking, false if it is done
- `obj.isDoneBaking()`: I'll let you fuckin guess lmao

`InventoryTray`
- `obj.isStale()`: Guess.

### CODE REVIEW

File: Core/src/main/java/io/htmlcss/model/BakingTray.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    public BakingTray(int donutID, int quantity, LocalDateTime startBakingTime) {
+        // End baking time is 20 minutes after start baking time
+        super(-1, donutID, quantity);
+        this.startBakingTime = startBakingTime;
+        endBakingTime = startBakingTime.plusMinutes(20);
+
+        // Insert self into the database to get a trayID
+        this.trayID = DBFactory.getDatabaseFetcher().insertTray(this);
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/BakingTray.java:26]: Creates the tray and automatically adds it to the database at creation.

File: Core/src/main/java/io/htmlcss/model/BakingTray.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+ * This object is a model object for passing or storing a baking tray object.
+ * The database will return this model.
+ */
+public class BakingTray extends Tray {
+    private LocalDateTime startBakingTime;
+    private LocalDateTime endBakingTime;
+
+    public BakingTray(Donut d, int quantity, LocalDateTime startBakingTime) {
+        // End baking time is 20 minutes after start baking time
+        this(d.getId(), quantity, startBakingTime);
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/BakingTray.java:16]: This means that all constructors add the tray.

File: Core/src/main/java/io/htmlcss/model/InventoryTray.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import java.time.LocalDateTime;
+
+/**
+ * This object is a model object for passing or storing an inventory tray object.
+ * The database will return this model.
+ * 
+ * InventoryTrays are automatically created when a BakingTray finishes baking.
+ * These constructors should not be called by other classes.
+ */
+public class InventoryTray extends Tray {
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/InventoryTray.java:11]: I will not create an inventory tray, that is for the Tray to handle.

File: Core/src/main/java/io/htmlcss/model/Tray.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+import io.htmlcss.db.*;
+
+import java.util.ArrayList;
+
+/**
+ * This object is a model object for passing or storing a tray object.
+ * The database will return this model.
+ */
+public abstract class Tray {
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/Tray.java:13]: This class creates an interface between the front end and the backend.

File: Core/src/main/java/io/htmlcss/model/Tray.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+     */
+    protected static void addTray(Tray tray) {
+        prefetchTrays();
+        trays.add(tray);
+    }
+
+    /**
+     * Get a list of trays that are currently being used for baking.
+     */
+    public static List<Tray> getBakingTrays() {
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/Tray.java:61]: This function will let me get all the trays that are currently in the oven.

File: Core/src/main/java/io/htmlcss/model/Tray.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            }
+        }
+
+        return bakingTrays;
+    }
+
+    /**
+     * Get a list of trays that are in the current inventory AND not stale.
+     */
+    public static List<Tray> getInventoryTrays() {
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/Tray.java:95]: Gets the trays of what donuts I do have to sell. This should be referred to by the checkout system.

File: Core/src/main/java/io/htmlcss/model/Tray.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                inventoryTrays.add(tray);
+            }
+        }
+        return inventoryTrays;
+    }
+
+    /**
+     * Get a list of trays that have gone stale.
+     */
+    public static List<Tray> getStaleTrays() {
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/Tray.java:111]: Korbin's implementation of getting stale tray information. Will be used by the admin report if Kassidy's implementation isn't used.

File: Core/src/main/java/io/htmlcss/model/Tray.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+        return bakedTrays;
+    }
+
+    /**
+     * Update the list of trays when the trays in the inventory have gone stale.
+     * This should be called after the baker has thrown away the stale trays.
+     * @return List of trays that have gone stale and should be removed from the inventory
+     */
+    public static List<Tray> updateStaleTrays() {
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/Tray.java:157]: Do not call function, to preserve information for stale report.

File: Core/src/main/java/io/htmlcss/model/Tray.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        }
+        return staleTrays;
+    }
+
+    /**
+     * Update the list of trays when the trays in the oven are done baking. (BakingTray -> InventoryTray)
+     * This should be called after the baker has removed the trays from the oven.
+     * @return List of trays that have finished baking and should be moved to the inventory
+     */
+    public static List<Tray> updateBakedTrays() {
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/Tray.java:129]: Create interface to let baker update the trays once he/she has pulled them out.

File: Core/src/main/java/io/htmlcss/model/Tray.java [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+            }
+        }
+
+        return bakingTrays;
+    }
+
+    /**
+     * Get a list of trays that have finished baking.
+     */
+    public static List<Tray> getBakedTrays() {
```
Comment (HeartAttackCat) [inline Core/src/main/java/io/htmlcss/model/Tray.java:78]: Trays that need to be pulled out. PULL YOUR DONUTS OUT OR ELSE 🔪🔪🔪

**Review by HeartAttackCat — COMMENTED** [review]
Overall use is

### BakingTray.java
- Create baking trays when employee adds them to the oven.
- Baking trays automatically add themselves to the database and be managed by Tray
### InventoryTray.java
- Will not initialize them.
### Tray.java
- `GetBakingTrays()` will get all trays in the oven.
- `GetBakedTrays()` will get trays that are baked.
- `getInventoryTrays()` will get trays of current stock.
- 'getStaleTrays()' Korbins implementation of getting stale trays.
- `updateBakedTrays()` Update the status of trays in the oven.
- `updateStaleTrays()` AVOID THIS WILL RESULT IN ISSUES WITH REPORT INFORMATION.

**Review by benimaru-afk — APPROVED** [review]
added new imports in 'DatabaseFetcher' 
added `tray` functions for adding, deleting, updating
added `BakingTray`
added

**Pull request merged**